keel-core/src/test/kotlin/com/netflix/spinnaker/keel/serialization/SubmittedEnvironmentDeserializerTests.kt

```diff
@@ -0,0 +1,175 @@
+package com.netflix.spinnaker.keel.serialization
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.jsontype.NamedType
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.SubmittedEnvironment
+import com.netflix.spinnaker.keel.api.SubnetAwareLocations
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import strikt.api.expectThat
+import strikt.api.expectThrows
+import strikt.assertions.isA
+import strikt.assertions.isEqualTo
+
+class SubmittedEnvironmentDeserializerTests : JUnit5Minutests {
+
+  data class Fixture(
+    val json: String
+  ) {
+    val mapper = configuredYamlMapper()
+      .apply {
+        configure(DeserializationFeature.WRAP_EXCEPTIONS, true)
+        registerSubtypes(NamedType(TestLocatableResource::class.java, "test/v1/test-locatable"))
+      }
+  }
+
+  fun tests() = rootContext<Fixture> {
+    context("locations is specified on the resource spec") {
+      fixture {
+        Fixture("""
+          |---
+          |name: test
+          |resources:
+          |- apiVersion: test/v1
+          |  kind: test-locatable
+          |  metadata:
+          |    serviceAccount: mickey@disney.com
+          |  spec:
+          |    id: my-resource
+          |    application: whatever
+          |    locations:
+          |      account: prod
+          |      subnet: internal (vpc0)
+          |      vpc: vpc0
+          |      regions:
+          |      - name: us-east-1
+          |      - name: us-west-2
+          |locations:
+          |  account: test
+          |  subnet: internal (vpc0)
+          |  vpc: vpc0
+          |  regions:
+          |  - name: us-east-1
+          |  - name: us-west-2
+          """.trimMargin()
+        )
+      }
+
+      test("locations corresponds to what was specified on the resource") {
+        val environment = mapper.readValue<SubmittedEnvironment>(json)
+
+        expectThat(environment.resources.first().spec)
+          .isA<TestLocatableResource>()
+          .get { locations.account }
+          .isEqualTo("prod")
+      }
+    }
+
+    context("locations is omitted from the resource spec") {
+      fixture {
+        Fixture("""
+          |---
+          |name: test
+          |resources:
+          |- apiVersion: test/v1
+          |  kind: test-locatable
+          |  metadata:
+          |    serviceAccount: mickey@disney.com
+          |  spec:
+          |    id: my-resource
+          |    application: whatever
+          |locations:
+          |  account: test
+          |  subnet: internal (vpc0)
+          |  vpc: vpc0
+          |  regions:
+          |  - name: us-east-1
+          |  - name: us-west-2
+          """.trimMargin()
+        )
+      }
+
+      test("locations corresponds to what was specified on the resource") {
```

Comment:
“...to what was specified on the parent environment”?

Reply:
Good catch. Bit of copy 'n paste there.